### PR TITLE
Bug in ChartManifold (SphericalManifold as an example)

### PR DESCRIPTION
**Pull request by luca-heltai**
@bangerth

This bug is preventing mapping-manifold from working. I'm trying a fix, but I don't really understand why it is not working...

This is what's happening. If I create a face quadrature, and use `get_new_point` based on the face vertices, for each face, then I get the following nice picture.

face_quads

If instead, I project the face quadrature to the cell (this is what happens inside `Mapping`), the `get_new_point` is used with the vertices of the **cell**, not the vertices of the **faces**. Of course, the weights associated to opposite face vertices are always zero. However, the periodicity trick in this case fails (I don't understand why), and for the face that crosses the periodic boundary (the lower one), we go the longer way (i.e., as if the manifold was not periodic across 2*pi). Using the projected quadrature ends up with the following:

cell_quads

where the quadrature points on the lower face get distributed along the longer circular arc, ending up somewhere mixed with the other faces.

### CODE REVIEW

File: tests/manifold/spherical_manifold_07.cc [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+      for(unsigned int v=0; v<weights.size(); ++v)
+	weights[v] = face_fe_q.shape_value(v, face_quad.point(i));
+
+      deallog << manifold.get_new_point(Quadrature<spacedim>(vertices, weights))
+	      << std::endl;
+    }
+  }
+
+  // Project the face quads
+  Quadrature<dim> quad = QProjector<dim>::project_to_all_faces(face_quad);
```
Comment (bangerth) [inline tests/manifold/spherical_manifold_07.cc]: This projects the quadrature points onto all four faces. For your QGauss formula, this results in 3*4=12 quadrature points.

File: tests/manifold/spherical_manifold_07.cc [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  std::vector<double> weights(GeometryInfo<dim>::vertices_per_cell);
+  
+  for (unsigned int i=0; i<quad.size(); ++i)
+    {
+      for(unsigned int v=0; v<weights.size(); ++v)
+	weights[v] = fe_q.shape_value(v, quad.point(i));
+
+      deallog << manifold.get_new_point(Quadrature<spacedim>(vertices, weights))
+	      << std::endl;
+    }
```
Comment (bangerth) [inline tests/manifold/spherical_manifold_07.cc]: So I'm not sure what exactly you do here. You loop over all 12 points along the perimeter of the reference cell, then compute for each of them the 4 weights, and compute the new-point operation for each of the points with the 4 vertices and 4 weights. I would expect this to yield 12 points on the perimeter of the *mapped* cell. Is this what you would expect as well?

**Comment by bangerth** [comment]
So is this a case where we violate the condition that it must be possible to cover all points given to the new-point operation with a single chart?

File: tests/manifold/spherical_manifold_07.cc (hunk 1 above)
Comment (luca-heltai) [inline tests/manifold/spherical_manifold_07.cc]: Above I loop over all faces, and project each quadrature on each face... I do manually what is done in QProjector

File: tests/manifold/spherical_manifold_07.cc (hunk 2 above)
Comment (luca-heltai) [inline tests/manifold/spherical_manifold_07.cc]: Yes. This is how things would be computed inside MappingManifold...

**Comment by luca-heltai** [comment]
I've added yet another test, which is simpler to understand. If I ask for a point between two surrounding points, using `w`, and `(1-w)` as weights, the result should be same as asking for a point between N>2 surrounding points, if all additional points with index > 1 have associated weight = 0. 

This has nothing to do with covering with a single chart. I think it is an issue with periodicity, which works ok for two points, but not for more than two, even if the additional points have a zero weight associated with them.

**Comment by luca-heltai** [comment]
The same test as above works as expected in `FlatManifold` but not in `SphericalManifold`. I'm trying to figure out what's wrong...

**Comment by bangerth** [comment]
I'm trying to look at this, but got confused by the sequence of commits. The flat_manifold_07, chart_manifold_09 tests seem to work as expected. I may misinterpret the output of spherical_manifold_07 but all points seem to lie on the entire circle. I haven't tried to understand the spherical_manifold_08 test.

Can you tell me again which test it is that doesn't work as you expect?

**Comment by luca-heltai** [comment]
Both `spherical_manifold_07` and `spherical_manifold_08` fail. The corresponding commits contain the expected output. I thought originally that the problem might have been in chart_manifold, but chart manifold works (that's why I added both `FlatManifold` and `ChartManifold` tests).

The bug is the following: take two surrounding points which cross the periodicity for a spherical manifold. Compute points between them by making a quadrature with `w` and `1-w`. Spherical manifold in this case correctly construct the section of arc between the two points crossing the periodicity boundary. 

Add two surrdoungin points with zero weight, and suddenly spherical manifold goes the other way (the long way) causing the two tests to fail. 

We never saw this, because we construct new points on faces using the vertices of the faces (2 in 2d). In Mapping+FEvalues (which is what I was trying to implement when I discovered the bug), however, the way FEFaceValues works, is to construct a **cell** quadrature, that contains all quadrature points of the face quadrature. This means: in 2d we construct intermediate points using 4 surrounding points, 2 of which always have a zero weight associated to it.

**Comment by bangerth** [comment]
I think I know what's happening here, but I don't know how to fix it. It is ultimately a problem of finding the chart that covers all points, in view of the periodicity. I think the fact that it works for the original face projection (with only 2 points, both of which have nonzero weights) is actually just an accident and one should be able to construct a failing test case by rotating the original square somehow.

My take of what happens is that we end up in
```
template <int dim, int spacedim>
Point<spacedim>
SphericalManifold<dim,spacedim>::get_new_point(const Quadrature<spacedim> &quad) const
{
  if (spacedim == 2)
    return ChartManifold<dim,spacedim,spacedim>::get_new_point(quad);
```
which calls this function:
```
template <int dim, int spacedim, int chartdim>
Point<spacedim>
ChartManifold<dim,spacedim,chartdim>::
get_new_point (const Quadrature<spacedim> &quad) const
{
  const std::vector<Point<spacedim> > &surrounding_points = quad.get_points();
  const std::vector<double> &weights = quad.get_weights();
  std::vector<Point<chartdim> > chart_points(surrounding_points.size());

  for (unsigned int i=0; i<surrounding_points.size(); ++i)
    chart_points[i] = pull_back(surrounding_points[i]);                  // *****************************

  const Quadrature<chartdim> chart_quad(chart_points, weights);
  const Point<chartdim> p_chart = sub_manifold.get_new_point(chart_quad);

  return push_forward(p_chart);
}
```
Now note that in the line marked with the asterisks, we pull back each of the points separately. The pull back is defined as follows:
```
template <int dim, int spacedim>
Point<spacedim>
SphericalManifold<dim,spacedim>::pull_back(const Point<spacedim> &space_point) const
{
  const Tensor<1,spacedim> R = space_point-center;
  const double rho = R.norm();

  Point<spacedim> p;
  p[0] = rho;

  switch (spacedim)
    {
    case 2:
      p[1] = atan2(R[1],R[0]);
      if (p[1] < 0)                                          // *****************************
        p[1] += 2*numbers::PI;                     // *****************************
      break;
```
For the given testcase, for *some*, but not *all* points we add 2*pi. I think this is still ok.

With these points we ultimately get here:
```
template <int dim, int spacedim>
Point<spacedim>
FlatManifold<dim, spacedim>::
get_new_point (const Quadrature<spacedim> &quad) const
{
  Assert(std::abs(std::accumulate(quad.get_weights().begin(), quad.get_weights().end(), 0.0)-1.0) < 1e-10,
         ExcMessage("The weights for the individual points should sum to 1!"));

  const std::vector<Point<spacedim> > &surrounding_points = quad.get_points();
  const std::vector<double> &weights = quad.get_weights();

  Point<spacedim> p;
  Point<spacedim> dp;
  Point<spacedim> minP = periodicity;
  const bool check_period = (periodicity.norm() > tolerance);
  if (check_period)
    for (unsigned int i=0; i<surrounding_points.size(); ++i)
      for (unsigned int d=0; d<spacedim; ++d)
        {
          minP[d] = std::min(minP[d], surrounding_points[i][d]);
          if (periodicity[d] > 0)
            Assert( (surrounding_points[i][d] < periodicity[d]+tolerance*periodicity.norm()) ||
                    (surrounding_points[i][d] >= -tolerance*periodicity.norm()),
                    ExcPeriodicBox(d, surrounding_points[i], periodicity, tolerance*periodicity.norm()));
        }

  for (unsigned int i=0; i<surrounding_points.size(); ++i)
    {
      dp = Point<spacedim>();
      if (check_period)
        {
          for (unsigned int d=0; d<spacedim; ++d)
            if (periodicity[d] > 0)
              dp[d] = ( (surrounding_points[i][d]-minP[d]) > periodicity[d]/2.0 ?
                        -periodicity[d] : 0.0 );
        }
      p += (surrounding_points[i]+dp)*weights[i];
    }
  if (check_period)
    for (unsigned int d=0; d<spacedim; ++d)
      if (periodicity[d] > 0)
        p[d] = (p[d] < 0 ? p[d] + periodicity[d] : p[d]);

  return project_to_manifold(surrounding_points, p);
}
```
where we now compute `minP` for all four, not just two points, and the result is different than for the 2-point case. I suppose this then causes the problem, but need to look into this some further later today.

I find the last of the functions quoted above a bit complicated to understand. I may rewrite it to see whether that helps anything. I think it would also help to reduce the _07 testcase to not use cells and quadratures, but to just create the point and weight arrays for the first face in which you get different results, and then call the spherical manifold directly. This would leave just two calls to `new_point` with fixed and known arguments, and one could debug the problem by going side-by-side stepping into the same function from the two call sites, then comparing the internal state.

I'll do that as well later today if you don't beat me to it.

**Comment by luca-heltai** [comment]
So, the minimal tests are the ones I already added. `spherical_manifold_08.cc` does trigger the bug without any additional information from a grid. Just using 2 points on round 0 and 4 points on round 1. The same should be true for the last chart manifold test. That test I added precisely to check your last assertion, that getting the min of four points does not yeld the same result of taking the min on the two points, but I was out of luck...

**Comment by bangerth** [comment]
OK, I should have looked at the _08 test. Will get back to you later when looking at that one.

**Comment by bangerth** [comment]
OK, so looking at this I've come to realize that I don't actually know what to do about this situation. Conceptually, you have a few points on the periodic segment `[0,2pi]`(the radius of these points is immaterial here) and you are looking for a new point in the middle with certain weights. *Implicitly*, what we try to do is to find the shortest interval on this periodic segment that covers all of these points (which may wrap around the end points) and then find the "average" within this shortest interval. This shortest interval corresponds to the *chart* in which we try to find the averaged point.

This can already go wrong if the two points are exactly pi apart, say at angles 0 and pi -- there are two such shortest intervals `[0,pi]` and `[pi,2pi]`, and one could with equal right claim that the new midpoint should be either pi/2 or 3pi/2. So this is not necessarily a problem with having 4 points, or having some with weight zero.

It also goes wrong for the case you describe. In essence, if you only have two points at 0 and pi/2 (allow me to rotate your square by 45 degrees), then it's clear what the shortest interval is, and the new point should be at pi/4. But if you add the other two points at pi and 3pi/2, you get into a situation where there are four smallest charts that are all equally valid: `[0,3pi/2]`, `[pi/2,2pi]`, `[pi,pi/2]`, `[3pi/2,pi]`. All of these have length 3pi/2, and depending on which one you fancy, you will get a different new weighted mid-point. I say "fancy" because I cannot come up with a criterion that would usefully choose one over the other.

So I'm genuinely stumped. All of this can only happen if we are in a periodic domain and points are spread out over more than (or equal to) one half of a period, which is the case in the situation above. This is likely not going to happen once you refine the mesh a number of times, but it can of course happen if you try to do the first refinement step.

I really don't know what to do here. What do you think?

**Comment by luca-heltai** [comment]
I think I solved the issue, by simply not using the information of the surrounding points if the corresponding weight is equal to zero. I'm not sure this is a good fix, but at least it makes things work for the cases I stumbled upon, and does not break anything else. 

I think this test case is really a corner case. We never encounter it anyway, because we **should not** use a square as a circle (and we don't, in general). Our circles are always made of 5 cells. And there nothing goes wrong, because the points never span the entire [0,2pi] interval.

On the other hand, if the surrounding points **do** span the entire [0,2pi] interval, then it *is* correct to take the long way, because it is true that the intermediate points are on the long side (as in this image:

image

If, however, the weights are zero, then it should be as if those points were never inserted in the Quadrature, and they should not be used when computing the minimum point, as in this (I think correct) cas:

image

This is clearly a corner case, and it reflects the fact that the mapping we are considering is not invertible in the domain we are considering. I don't think we can do anything better w.r.t. what we already do here... 

This issue points out some important aspects of the current `Manifold` design, that are worth mentioning...

**Comment by bangerth** [comment]
On 03/26/2016 04:58 AM, Luca Heltai wrote:
> I think I solved the issue, by simply not using the information of the
> surrounding points if the corresponding weight is equal to zero. I'm not sure
> this is a good fix, but at least it makes things work for the cases I stumbled
> upon, and does not break anything else.

I'm not a fan of this. We should make sure that the returned point is a 
continuous function of the input points and the weights. Checking for whether 
a weight is zero is going to ask for trouble.


> I think this test case is really a corner case. We never encounter it anyway,
> because we *should not* use a square as a circle (and we don't, in general).
> Our circles are always made of 5 cells. And there nothing goes wrong, because
> the points never span the entire [0,2pi] interval.

Correct. We could argue that we simply don't want to fix this situation.


> On the other hand, if the surrounding points *do* span the entire [0,2pi]
> interval, then it /is/ correct to take the long way, because it is true that
> the intermediate points are on the long side (as in this image:

Yes. The question is how many points you have along that "long" segment. As I 
mentioned in my previous comment, I think what we try to do is find the 
"minimal" interval to cover all points. If you only have 4 points around the 
circle, there are a number of equally valid options. But if you have 10 points 
equally spaced 3/4 around the circle, you are right that there is only one way 
to define the "minimal" segment, and going around the long way is the only 
reasonable choice.


> This issue points out some important aspects of the current |Manifold| design,
> that are worth mentioning...

Yes, indeed, this should go into the paper. I'll add a note.

**Comment by bangerth** [comment]
You merged this into your own repository, but I'd be interested in having these testcases also on the dealii master -- though I'm not a fan of the exclusion of points with zero weights. What do you think?

**Comment by luca-heltai** [comment]
If you wish, I could make a PR to master. However, I hope the other PR  will be ready soon, as this was one of the last remaining obstacles for MappingManifold to work. Nobody ever stumbled across this, since we don't yet have a Mapping that computes `FEFaceValues`...

**Comment by luca-heltai** [comment]
As far as the zero weight is concerned, I argue in the following way: consider always pairwise points:

- if the second weight is zero, the resulting point is the first one (and the other way around)
- with the resulting point, do the same with the next point. If the weight is zero, then we leave the point as is
- proceed until you finished all points. 

The only reason why things were not working is because our algorithm internally tries to be smart and it uses all points at once, while in reality it should do only pair-wise computations (as you wrote in our draft). With pairwise computations, things work correctly, with one exception, i.e., where the distance is **exactly** half a period, in which case the problem is not univocally solvable as we could in principle go both ways (there the algorithm will make one choice based on truncation, I assume). This we should not fix. We probably should discuss it in depth in our draft,  and for sure we should document it better in the code... But it is not broken.

**Comment by bangerth** [comment]
Well, I would like to have the testcases. But I don't like the change that excludes the points with small weights. I think this is asking for trouble. I would just declare that the current output of these tests is correct.

**Comment by bangerth** [comment]
Regarding the zero weights: that's actually quite an interesting point of view. I think you convinced me. Thanks for the insight -- I will address this in more detail in the paper.

So do you want to update https://github.com/dealii/dealii/pull/2420 after the fact?

**Comment by bangerth** [comment]
I see that the zero weight issue is addressed in the `MappingManifold` patch.
